frontend_vm/src/compiler.rs

```diff
@@ -0,0 +1,958 @@
+                        // so we do nothing.
+                    }
+                    DataValue::Direct(bytes) => {
+                        for i in (0..bytes.len()).step_by(4) {
+                            let v = (0..4)
+                                .map(|j| {
+                                    (bytes.get(i + j).cloned().unwrap_or_default() as u32)
+                                        << (j * 8)
+                                })
+                                .reduce(|a, b| a | b)
+                                .unwrap();
+                            // We can assume memory to be zero-initialized.
+                            if v != 0 {
+                                object_code.extend([
+                                    format!("addr <=X= 0x{:x};", pos + i as u32),
+                                    format!("mstore 0x{v:x};"),
+                                ]);
+                            }
+                        }
+                    }
+                    DataValue::Reference(sym) => {
+                        object_code.push(format!("addr <=X= 0x{pos:x};"));
+                        if let Some(p) = positions.get(sym) {
+                            object_code.push(format!("mstore 0x{p:x};"));
+                        } else {
+                            // code reference
+                            // TODO should be possible without temporary
+                            object_code.extend([
+                                format!("tmp1 <=X= load_label({});", escape_label(sym)),
+                                "mstore tmp1;".to_string(),
+                            ]);
+                        }
+                    }
+                }
+                pos += item.size() as u32;
+            }
+            if let Some(first_line) = object_code.first_mut() {
+                *first_line = format!("// data {name}\n") + first_line;
+            }
+            object_code
+        })
+        .collect();
+    (code, positions)
+}
+
+fn next_multiple_of_four(x: usize) -> usize {
+    ((x + 3) / 4) * 4
+}
+
+pub fn substitute_symbols_with_values(
+    mut statements: Vec<Statement>,
+    data_positions: &BTreeMap<String, u32>,
+) -> Vec<Statement> {
+    for s in &mut statements {
+        let Statement::Instruction(_name, args) = s else { continue; };
+        for arg in args {
+            arg.post_visit_expressions_mut(&mut |expression| match expression {
+                Expression::Number(_) => {}
+                Expression::Symbol(symb) => {
+                    if let Some(pos) = data_positions.get(symb) {
+                        *expression = Expression::Number(*pos as i64)
+                    }
+                }
+                Expression::UnaryOp(op, subexpr) => {
+                    if let Expression::Number(num) = subexpr[0] {
+                        let result = match op {
+                            UnaryOpKind::HiDataRef => num >> 12,
```

Comment:
I think HiDataRef is also riscv-specific, at least the number of bits that are considered to be "low bits".